Framework/PythonInterface/mantid/api/src/Exports/Sample.cpp

```diff
@@ -81,5 +82,8 @@ void export_Sample() {
       .def("__len__", &Sample::size, arg("self"),
            "Gets the number of samples in this collection")
       .def("__getitem__", &Sample::operator[],(arg("self"), arg("index")),
-           return_internal_reference<>());
+           return_internal_reference<>())
+      .def("__copy__", &Mantid::PythonInterface::generic__copy__<Sample>)
+      .def("__deepcopy__",
+           &Mantid::PythonInterface::generic__deepcopy__<Sample>);
```

Comment:
Can you elaborate on the reason why we need this heavy machinery here, instead of just returning a plain copy in both cases?

Reply:
If we try to return a  copy by value then we end up with 2 copies. The one in that we copy and return and then one inside boost.python (https://github.com/boostorg/python/blob/develop/include/boost/python/object/value_holder.hpp#L132). Boost Python constructs this internal value holder and with value types it has to copy them but with heap allocated types it can just push in the pointer.

So we are doing the dance to allocate on the heap using the copy constructor and then push that pointer to the boost.python object holder.

Reply:
@martyngigg I didn't realise that there were 2 copies as a problem. I did consider changing the return_value_policy from [this](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/api/src/Exports/ExperimentInfo.cpp#L80) to `copy_const_reference` and then remove the `noncopyable` from `Sample`. The reason I opted for the approach used here was that it allows clients to make further copies if needed.

Reply:
@martyngigg last commit does this for Sample. Applying the same to `Run` causes test failures. I suspect this has something to do with copying the `LogManager` base.